compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/KernelDispatch.cpp

```diff
@@ -486,20 +489,30 @@ static LogicalResult setRootConfigForPackPeel4LevelTilingPipeline(
   // ------------------------------------------------------
   // -------------- Set lowering config -------------------
   // ------------------------------------------------------
-  // Check if we can scale L2 size of A and B with a factor of 2. TODO(jornt):
-  // generalize to find largest scaling factor possible.
+  // Check if we can scale L2 size of A and B with a factor.
+  // TODO(jornt): This assumes double buffering to get an upper bound, but
+  // retrieve this information instead.
   int64_t l2SizeA =
       2 * packPeelTiling.M0 * packPeelTiling.K * packPeelTiling.nBytesLhs;
   int64_t l2SizeB =
-      2 * packPeelTiling.N0 * packPeelTiling.K * packPeelTiling.nBytesRhs;
-  int64_t l2SizeInit =
-      4 * packPeelTiling.M0 * packPeelTiling.N0 * packPeelTiling.nBytesInit;
+      4 * packPeelTiling.N0 * packPeelTiling.K * packPeelTiling.nBytesRhs;
+  int64_t l2SizeOut =
+      4 * packPeelTiling.M0 * packPeelTiling.N0 * packPeelTiling.nBytesOut;
 
-  bool fitsInL2 = (l2SizeA + l2SizeB + l2SizeInit) <
+  bool fitsInL2 = (l2SizeA + l2SizeB + l2SizeOut) <
                   (deviceModel.getMemTileSizeInBytes() * numCols);
-  int64_t scaleL0 = !isBatchMatmul && fitsInL2 ? 2 : 1;
-  int64_t m0Tile = packPeelTiling.M0 * scaleL0;
-  int64_t n0Tile = packPeelTiling.N0 * scaleL0;
+  int64_t largestM0ScaleFactor =
+      ((deviceModel.getMemTileSizeInBytes() * numCols) - l2SizeB) /
+      (l2SizeA + l2SizeOut);
+  if (largestM0ScaleFactor % 2 != 0)
+    largestM0ScaleFactor = llvm::PowerOf2Ceil(largestM0ScaleFactor) / 2;
+  int64_t scaleL0M =
+      (isObjectFifo && !isBatchMatmul) ? largestM0ScaleFactor : 1;
+  int64_t scaleL0N = (isObjectFifo && !isBatchMatmul && fitsInL2) ? 2 : 1;
+  int64_t m0Tile = std::min(static_cast<uint32_t>(packPeelTiling.M0 * scaleL0M),
+                            packPeelTiling.M);
+  int64_t n0Tile = std::min(static_cast<uint32_t>(packPeelTiling.N0 * scaleL0N),
+                            packPeelTiling.N);
```

Comment:
I couldn't follow the logic here. Could you add some explanations? Also I don't think you need to check `isBatchMatmul` since the tile size for batch dimension in L2 is always 1.

Reply:
I added a comment for this. As for `isBatchMatmul`, this was added because batch matmul failed with 'pack-peel-4-level-tiling' and as I think enabling that is out-of-scope of this PR, so I added a TODO for it.